edward/inferences/renyi_divergence.py

```diff
@@ -0,0 +1,176 @@
+  [@li2016renyi].
+
+  #### Notes
+      + The gradient estimator used here does not have any analytic version.
+      + The gradient estimator used here does not have any version for non
+      reparametrizable models.
+      + backward_pass = 'max': (extreme case $\alpha \rightarrow -\infty$)
+      the algorithm chooses the sample that has the maximum unnormalised
+      importance weight. This does not minimize the Renyi divergence
+      anymore.
+      + backward_pass = 'min': (extreme case $\alpha \rightarrow +\infty$)
+      the algorithm chooses the sample that has the minimum unnormalised
+      importance weight. This does not minimize the Renyi divergence
+      anymore. This mode is not describe in the paper but implemented
+      in the publicly available implementation of the paper's experiments.
+  """
+
+  def __init__(self, *args, **kwargs):
+
+    super(RenyiDivergence, self).__init__(*args, **kwargs)
+
+    self.is_reparameterizable = all([
+        rv.reparameterization_type ==
+        tf.contrib.distributions.FULLY_REPARAMETERIZED
+        for rv in six.itervalues(self.latent_vars)])
+
+  def initialize(self,
+                 n_samples=32,
+                 alpha=1.0,
+                 backward_pass='full',
+                 *args, **kwargs):
+    """Initialize inference algorithm. It initializes hyperparameters
+    and builds ops for the algorithm's computation graph.
+
+    Args:
+        n_samples: int, optional.
+            Number of samples from variational model for calculating
+            stochastic gradients.
+        alpha: float, optional.
+            Renyi divergence coefficient. $\alpha \in \mathbb{R}$.
+            When $\alpha < 0$, the algorithm still does something sensible but
+            does not minimize the Renyi divergence anymore.
+            (see [@li2016renyi] - section 4.2)
+        backward_pass: str, optional.
+            Backward pass mode to be used.
+            Options: 'min', 'max', 'full'
+            (see [@li2016renyi] - section 4.2)
+    """
+    self.n_samples = n_samples
+    self.alpha = alpha
+    self.backward_pass = backward_pass
+
+    return super(RenyiDivergence, self).initialize(*args, **kwargs)
+
+  def build_loss_and_gradients(self, var_list):
+    """Build the Renyi ELBO function.
+
+    Its automatic differentiation is a stochastic gradient of
+
+    $ \mcalL_{R}^{\alpha}(q; x) =
+            \frac{1}{1-\alpha} \log \dsE_{q} \left[
+                \left( \frac{p(x, z)}{q(z)}\right)^{1-\alpha} \right].$
+
+    It uses:
+        + Monte Carlo approximation of the ELBO [@li2016renyi].
+        + Reparameterization gradients [@kingma2014auto].
+        + Stochastic approximation of the joint distribution [@li2016renyi].
+
+    #### Notes
+        + If the model is not reparameterizable, it returns a
+        NotImplementedError.
+        + See Renyi Divergence Variational Inference [@li2016renyi] for
+        more details.
+    """
+    if self.is_reparameterizable:
```

Comment:
`is_reparameterizable` should be checked with the possible raising error during the `__init__`, and since it's checked there it doesn't need to be stored in the class. This also helps to remove one layer of indentation in this function.

Reply:
Done